File: backend/core-service/src/models/Conversation.js
const { DataTypes } = require('sequelize');
const { sequelize } = require('../config/database');

const Conversation = sequelize.define('Conversation', {
  id: {
    type: DataTypes.UUID,
    defaultValue: DataTypes.UUIDV4,
    primaryKey: true,
  },
  session_id: {
    type: DataTypes.STRING(64),
    allowNull: false,
    comment: '会话ID',
  },
  user_id: {
    type: DataTypes.UUID,
    allowNull: true,
    comment: '用户ID（可为空，支持匿名会话）',
  },
  knowledge_base_id: {
    type: DataTypes.UUID,
    allowNull: true,
    comment: '使用的知识库ID',
  },
  title: {
    type: DataTypes.STRING(500),
    allowNull: true,
    comment: '对话标题',
  },
  type: {
    type: DataTypes.ENUM('chat', 'search', 'qa', 'support'),
    defaultValue: 'chat',
    allowNull: false,
    comment: '对话类型',
  },
  status: {
    type: DataTypes.ENUM('active', 'ended', 'archived'),
    defaultValue: 'active',
    allowNull: false,
    comment: '对话状态',
  },
  messages: {
    type: DataTypes.JSONB,
    defaultValue: [],
    comment: '消息列表',
  },
  context: {
    type: DataTypes.JSONB,
    defaultValue: {},
    comment: '对话上下文',
  },
  metadata: {
    type: DataTypes.JSONB,
    defaultValue: {},
    comment: '扩展元数据',
  },
  settings: {
    type: DataTypes.JSONB,
    defaultValue: {},
    comment: '对话设置',
  },
  tags: {
    type: DataTypes.ARRAY(DataTypes.STRING),
    defaultValue: [],
    comment: '标签数组',
  },
  rating: {
    type: DataTypes.INTEGER,
    allowNull: true,
    comment: '用户评分（1-5）',
  },
  feedback: {
    type: DataTypes.TEXT,
    allowNull: true,
    comment: '用户反馈',
  },
  message_count: {
    type: DataTypes.INTEGER,
    defaultValue: 0,
    comment: '消息数量',
  },
  last_message_at: {
    type: DataTypes.DATE,
    allowNull: true,
    comment: '最后消息时间',
  },
  started_at: {
    type: DataTypes.DATE,
    defaultValue: DataTypes.NOW,
    comment: '对话开始时间',
  },
  ended_at: {
    type: DataTypes.DATE,
    allowNull: true,
    comment: '对话结束时间',
  },
  client_info: {
    type: DataTypes.JSONB,
    defaultValue: {},
    comment: '客户端信息',
  },
  ip_address: {
    type: DataTypes.STRING(45),
    allowNull: true,
    comment: 'IP地址',
  },
  user_agent: {
    type: DataTypes.STRING(1000),
    allowNull: true,
    comment: '用户代理',
  },
}, {
  tableName: 'conversations',
  timestamps: true,
  underscored: true,
  indexes: [
    {
      fields: ['session_id'],
      unique: true,
      name: 'idx_conversations_session_id',
    },
    {
      fields: ['user_id'],
      name: 'idx_conversations_user_id',
    },
    {
      fields: ['knowledge_base_id'],
      name: 'idx_conversations_knowledge_base_id',
    },
    {
      fields: ['type'],
      name: 'idx_conversations_type',
    },
    {
      fields: ['status'],
      name: 'idx_conversations_status',
    },
    {
      fields: ['started_at'],
      name: 'idx_conversations_started_at',
    },
    {
      fields: ['last_message_at'],
      name: 'idx_conversations_last_message_at',
    },
    {
      fields: ['tags'],
      using: 'gin',
      name: 'idx_conversations_tags',
    },
    {
      fields: ['rating'],
      name: 'idx_conversations_rating',
    },
  ],
  hooks: {
    beforeCreate: (conversation) => {
      if (!conversation.title && conversation.messages.length > 0) {
        // 自动生成标题（使用第一条用户消息的前50个字符）
        const firstUserMessage = conversation.messages.find(msg => msg.role === 'user');
        if (firstUserMessage) {
          conversation.title = firstUserMessage.content.substring(0, 50) + 
            (firstUserMessage.content.length > 50 ? '...' : '');
        }
      }
    },
    beforeUpdate: (conversation) => {
      if (conversation.changed('messages')) {
        conversation.message_count = conversation.messages.length;
        if (conversation.messages.length > 0) {
          const lastMessage = conversation.messages[conversation.messages.length - 1];
          conversation.last_message_at = new Date(lastMessage.timestamp);
        }
      }
    },
  },
});

// 消息模型（嵌套在对话中）
const MessageSchema = {
  id: 'string', // UUID
  role: 'string', // 'user', 'assistant', 'system'
  content: 'string', // 消息内容
  content_type: 'string', // 'text', 'image', 'file', 'audio'
  timestamp: 'string', // ISO日期字符串
  metadata: 'object', // 扩展数据
  sources: 'array', // 来源文档
  tokens: 'object', // token使用统计
  processing_time: 'number', // 处理时间（毫秒）
  error: 'object', // 错误信息（如果有）
};

// 实例方法
Conversation.prototype.toJSON = function() {
  const values = { ...this.dataValues };
  
  // 格式化日期
  ['created_at', 'updated_at', 'started_at', 'ended_at', 'last_message_at'].forEach(field => {
    if (values[field]) {
      values[field] = values[field].toISOString();
    }
  });
  
  return values;
};

Conversation.prototype.addMessage = async function(message) {
  const messageWithId = {
    id: require('uuid').v4(),
    timestamp: new Date().toISOString(),
    ...message,
  };
  
  this.messages = [...this.messages, messageWithId];
  this.message_count = this.messages.length;
  this.last_message_at = new Date();
  
  await this.save();
  return messageWithId;
};

Conversation.prototype.updateLastMessage = async function(updates) {
  if (this.messages.length === 0) {
    throw new Error('No messages to update');
  }
  
  const lastMessageIndex = this.messages.length - 1;
  this.messages[lastMessageIndex] = {
    ...this.messages[lastMessageIndex],
    ...updates,
  };
  
  await this.save();
  return this.messages[lastMessageIndex];
};

Conversation.prototype.endConversation = async function(rating = null, feedback = null) {
  this.status = 'ended';
  this.ended_at = new Date();
  
  if (rating !== null) {
    this.rating = rating;
  }
  
  if (feedback !== null) {
    this.feedback = feedback;
  }
  
  await this.save();
};

Conversation.prototype.getMessages = function(limit = null, offset = 0) {
  const messages = this.messages || [];
  
  if (limit === null) {
    return messages.slice(offset);
  }
  
  return messages.slice(offset, offset + limit);
};

Conversation.prototype.getLastMessages = function(count = 10) {
  const messages = this.messages || [];
  return messages.slice(-count);
};

// 静态方法
Conversation.findBySessionId = function(sessionId) {
  return this.findOne({
    where: { session_id: sessionId },
  });
};

Conversation.findByUser = function(userId, options = {}) {
  return this.findAll({
    where: { user_id: userId },
    order: [['last_message_at', 'DESC']],
    ...options,
  });
};

Conversation.findActiveByUser = function(userId, options = {}) {
  return this.findAll({
    where: { 
      user_id: userId,
      status: 'active'
    },
    order: [['last_message_at', 'DESC']],
    ...options,
  });
};

Conversation.findByKnowledgeBase = function(knowledgeBaseId, options = {}) {
  return this.findAll({
    where: { knowledge_base_id: knowledgeBaseId },
    order: [['started_at', 'DESC']],
    ...options,
  });
};

Conversation.getStatsByPeriod = async function(startDate, endDate, knowledgeBaseId = null) {
  const { QueryTypes } = require('sequelize');
  
  let whereClause = 'WHERE started_at BETWEEN :startDate AND :endDate';
  const replacements = { startDate, endDate };
  
  if (knowledgeBaseId) {
    whereClause += ' AND knowledge_base_id = :knowledgeBaseId';
    replacements.knowledgeBaseId = knowledgeBaseId;
  }
  
  const query = `
    SELECT 
      COUNT(*) as total_conversations,
      AVG(message_count) as avg_message_count,
      AVG(rating) as avg_rating,
      COUNT(CASE WHEN status = 'ended' THEN 1 END) as completed_conversations,
      COUNT(CASE WHEN rating IS NOT NULL THEN 1 END) as rated_conversations
    FROM conversations 
    ${whereClause}
  `;
  
  const result = await sequelize.query(query, {
    replacements,
    type: QueryTypes.SELECT,
  });
  
  return result[0];
};

Conversation.findPopularTopics = async function(knowledgeBaseId = null, limit = 10) {
  const { QueryTypes } = require('sequelize');
  
  let whereClause = 'WHERE status = \'ended\' AND title IS NOT NULL';
  const replacements = { limit };
  
  if (knowledgeBaseId) {
    whereClause += ' AND knowledge_base_id = :knowledgeBaseId';
    replacements.knowledgeBaseId = knowledgeBaseId;
  }
  
  const query = `
    SELECT 
      title, 
      COUNT(*) as count,
      AVG(rating) as avg_rating,
      AVG(message_count) as avg_messages
    FROM conversations 
    ${whereClause}
    GROUP BY title
    ORDER BY count DESC
    LIMIT :limit
  `;
  
  return await sequelize.query(query, {
    replacements,
    type: QueryTypes.SELECT,
  });
};

module.exports = Conversation; 

File: mobile/lib/features/conversation/domain/entities/conversation.dart
import 'package:equatable/equatable.dart';

/// 对话消息实体
class ConversationMessage extends Equatable {
  const ConversationMessage({
    required this.id,
    required this.role,
    required this.content,
    required this.contentType,
    required this.timestamp,
    this.metadata = const {},
    this.sources = const [],
    this.tokens,
    this.processingTime,
    this.error,
  });

  final String id;
  final MessageRole role;
  final String content;
  final String contentType;
  final DateTime timestamp;
  final Map<String, dynamic> metadata;
  final List<MessageSource> sources;
  final TokenUsage? tokens;
  final int? processingTime;
  final MessageError? error;

  @override
  List<Object?> get props => [
        id,
        role,
        content,
        contentType,
        timestamp,
        metadata,
        sources,
        tokens,
        processingTime,
        error,
      ];

  ConversationMessage copyWith({
    String? id,
    MessageRole? role,
    String? content,
    String? contentType,
    DateTime? timestamp,
    Map<String, dynamic>? metadata,
    List<MessageSource>? sources,
    TokenUsage? tokens,
    int? processingTime,
    MessageError? error,
  }) {
    return ConversationMessage(
      id: id ?? this.id,
      role: role ?? this.role,
      content: content ?? this.content,
      contentType: contentType ?? this.contentType,
      timestamp: timestamp ?? this.timestamp,
      metadata: metadata ?? this.metadata,
      sources: sources ?? this.sources,
      tokens: tokens ?? this.tokens,
      processingTime: processingTime ?? this.processingTime,
      error: error ?? this.error,
    );
  }
}

/// 消息角色枚举
enum MessageRole {
  user,
  assistant,
  system;

  String get value {
    switch (this) {
      case MessageRole.user:
        return 'user';
      case MessageRole.assistant:
        return 'assistant';
      case MessageRole.system:
        return 'system';
    }
  }

  static MessageRole fromString(String value) {
    switch (value.toLowerCase()) {
      case 'user':
        return MessageRole.user;
      case 'assistant':
        return MessageRole.assistant;
      case 'system':
        return MessageRole.system;
      default:
        throw ArgumentError('Invalid message role: $value');
    }
  }
}

/// 消息来源
class MessageSource extends Equatable {
  const MessageSource({
    required this.type,
    required this.id,
    required this.title,
    required this.content,
    required this.similarity,
  });

  final String type;
  final String id;
  final String title;
  final String content;
  final double similarity;

  @override
  List<Object> get props => [type, id, title, content, similarity];
}

/// Token使用统计
class TokenUsage extends Equatable {
  const TokenUsage({
    required this.promptTokens,
    required this.completionTokens,
    required this.totalTokens,
  });

  final int promptTokens;
  final int completionTokens;
  final int totalTokens;

  @override
  List<Object> get props => [promptTokens, completionTokens, totalTokens];
}

/// 消息错误
class MessageError extends Equatable {
  const MessageError({
    required this.code,
    required this.message,
    this.details,
  });

  final String code;
  final String message;
  final Map<String, dynamic>? details;

  @override
  List<Object?> get props => [code, message, details];
}

/// 对话实体
class Conversation extends Equatable {
  const Conversation({
    required this.id,
    required this.sessionId,
    this.userId,
    this.knowledgeBaseId,
    this.title,
    required this.type,
    required this.status,
    required this.messages,
    this.context = const {},
    this.settings = const {},
    this.tags = const [],
    this.rating,
    this.feedback,
    required this.messageCount,
    this.lastMessageAt,
    required this.startedAt,
    this.endedAt,
    this.clientInfo = const {},
    this.ipAddress,
    this.userAgent,
    required this.createdAt,
    required this.updatedAt,
  });

  final String id;
  final String sessionId;
  final String? userId;
  final String? knowledgeBaseId;
  final String? title;
  final ConversationType type;
  final ConversationStatus status;
  final List<ConversationMessage> messages;
  final Map<String, dynamic> context;
  final Map<String, dynamic> settings;
  final List<String> tags;
  final int? rating;
  final String? feedback;
  final int messageCount;
  final DateTime? lastMessageAt;
  final DateTime startedAt;
  final DateTime? endedAt;
  final Map<String, dynamic> clientInfo;
  final String? ipAddress;
  final String? userAgent;
  final DateTime createdAt;
  final DateTime updatedAt;

  @override
  List<Object?> get props => [
        id,
        sessionId,
        userId,
        knowledgeBaseId,
        title,
        type,
        status,
        messages,
        context,
        settings,
        tags,
        rating,
        feedback,
        messageCount,
        lastMessageAt,
        startedAt,
        endedAt,
        clientInfo,
        ipAddress,
        userAgent,
        createdAt,
        updatedAt,
      ];

  Conversation copyWith({
    String? id,
    String? sessionId,
    String? userId,
    String? knowledgeBaseId,
    String? title,
    ConversationType? type,
    ConversationStatus? status,
    List<ConversationMessage>? messages,
    Map<String, dynamic>? context,
    Map<String, dynamic>? settings,
    List<String>? tags,
    int? rating,
    String? feedback,
    int? messageCount,
    DateTime? lastMessageAt,
    DateTime? startedAt,
    DateTime? endedAt,
    Map<String, dynamic>? clientInfo,
    String? ipAddress,
    String? userAgent,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Conversation(
      id: id ?? this.id,
      sessionId: sessionId ?? this.sessionId,
      userId: userId ?? this.userId,
      knowledgeBaseId: knowledgeBaseId ?? this.knowledgeBaseId,
      title: title ?? this.title,
      type: type ?? this.type,
      status: status ?? this.status,
      messages: messages ?? this.messages,
      context: context ?? this.context,
      settings: settings ?? this.settings,
      tags: tags ?? this.tags,
      rating: rating ?? this.rating,
      feedback: feedback ?? this.feedback,
      messageCount: messageCount ?? this.messageCount,
      lastMessageAt: lastMessageAt ?? this.lastMessageAt,
      startedAt: startedAt ?? this.startedAt,
      endedAt: endedAt ?? this.endedAt,
      clientInfo: clientInfo ?? this.clientInfo,
      ipAddress: ipAddress ?? this.ipAddress,
      userAgent: userAgent ?? this.userAgent,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }

  /// 获取最后一条消息
  ConversationMessage? get lastMessage {
    return messages.isNotEmpty ? messages.last : null;
  }

  /// 获取最后一条用户消息
  ConversationMessage? get lastUserMessage {
    return messages
        .where((message) => message.role == MessageRole.user)
        .lastOrNull;
  }

  /// 获取最后一条助手消息
  ConversationMessage? get lastAssistantMessage {
    return messages
        .where((message) => message.role == MessageRole.assistant)
        .lastOrNull;
  }

  /// 是否为活跃对话
  bool get isActive => status == ConversationStatus.active;

  /// 是否已结束
  bool get isEnded => status == ConversationStatus.ended;

  /// 是否已归档
  bool get isArchived => status == ConversationStatus.archived;

  /// 是否有评分
  bool get hasRating => rating != null;
}

/// 对话类型枚举
enum ConversationType {
  chat,
  search,
  qa,
  support;

  String get value {
    switch (this) {
      case ConversationType.chat:
        return 'chat';
      case ConversationType.search:
        return 'search';
      case ConversationType.qa:
        return 'qa';
      case ConversationType.support:
        return 'support';
    }
  }

  String get displayName {
    switch (this) {
      case ConversationType.chat:
        return '聊天';
      case ConversationType.search:
        return '搜索';
      case ConversationType.qa:
        return '问答';
      case ConversationType.support:
        return '支持';
    }
  }

  static ConversationType fromString(String value) {
    switch (value.toLowerCase()) {
      case 'chat':
        return ConversationType.chat;
      case 'search':
        return ConversationType.search;
      case 'qa':
        return ConversationType.qa;
      case 'support':
        return ConversationType.support;
      default:
        throw ArgumentError('Invalid conversation type: $value');
    }
  }
}

/// 对话状态枚举
enum ConversationStatus {
  active,
  ended,
  archived;

  String get value {
    switch (this) {
      case ConversationStatus.active:
        return 'active';
      case ConversationStatus.ended:
        return 'ended';
      case ConversationStatus.archived:
        return 'archived';
    }
  }

  String get displayName {
    switch (this) {
      case ConversationStatus.active:
        return '活跃';
      case ConversationStatus.ended:
        return '已结束';
      case ConversationStatus.archived:
        return '已归档';
    }
  }

  static ConversationStatus fromString(String value) {
    switch (value.toLowerCase()) {
      case 'active':
        return ConversationStatus.active;
      case 'ended':
        return ConversationStatus.ended;
      case 'archived':
        return ConversationStatus.archived;
      default:
        throw ArgumentError('Invalid conversation status: $value');
    }
  }
}

/// 对话统计
class ConversationStats extends Equatable {
  const ConversationStats({
    required this.overview,
    required this.breakdowns,
  });

  final ConversationOverview overview;
  final ConversationBreakdowns breakdowns;

  @override
  List<Object> get props => [overview, breakdowns];
}

/// 对话概览统计
class ConversationOverview extends Equatable {
  const ConversationOverview({
    required this.totalConversations,
    required this.activeConversations,
    required this.avgRating,
    required this.ratedCount,
    required this.totalMessages,
  });

  final int totalConversations;
  final int activeConversations;
  final double avgRating;
  final int ratedCount;
  final int totalMessages;

  @override
  List<Object> get props => [
        totalConversations,
        activeConversations,
        avgRating,
        ratedCount,
        totalMessages,
      ];
}

/// 对话分类统计
class ConversationBreakdowns extends Equatable {
  const ConversationBreakdowns({
    required this.byType,
    required this.byStatus,
  });

  final List<ConversationTypeStats> byType;
  final List<ConversationStatusStats> byStatus;

  @override
  List<Object> get props => [byType, byStatus];
}

/// 按类型统计
class ConversationTypeStats extends Equatable {
  const ConversationTypeStats({
    required this.type,
    required this.count,
  });

  final ConversationType type;
  final int count;

  @override
  List<Object> get props => [type, count];
}

/// 按状态统计
class ConversationStatusStats extends Equatable {
  const ConversationStatusStats({
    required this.status,
    required this.count,
  });

  final ConversationStatus status;
  final int count;

  @override
  List<Object> get props => [status, count];
}

/// 发送消息请求
class SendMessageRequest extends Equatable {
  const SendMessageRequest({
    required this.content,
    this.contentType = 'text',
    this.metadata = const {},
  });

  final String content;
  final String contentType;
  final Map<String, dynamic> metadata;

  @override
  List<Object> get props => [content, contentType, metadata];
}

/// 发送消息响应
class SendMessageResponse extends Equatable {
  const SendMessageResponse({
    required this.userMessage,
    required this.assistantMessage,
    required this.sources,
    required this.processingTime,
  });

  final ConversationMessage userMessage;
  final ConversationMessage assistantMessage;
  final List<MessageSource> sources;
  final int processingTime;

  @override
  List<Object> get props => [userMessage, assistantMessage, sources, processingTime];
}

extension on List<ConversationMessage> {
  ConversationMessage? get lastOrNull => isEmpty ? null : last;
} 